### PR TITLE
Port the cljs-tooling completion machinery to compliment

### PR DESCRIPTION
**Pull request by arichiardi**
This PR shows what CLJS completions look like.

The code is ported from `cljs-tooling` - what is used by Cider for ClojureScript completions.

The goal here is to end up with the same public api and behavior. Probably some
of the Compliment options would not apply in which case, please, advice :smile:

I am always open to tweaks of any sort, just send any feedback over.

### CODE REVIEW

**Comment by Olical** [comment]
I'm extremely excited about this and will be hooking it into my prepl neovim tooling immediately. If I get around to it quickly enough I'll try to provide feedback and improvements where applicable.

**Comment by arichiardi** [comment]
@Olical thank you! Note that this isn't working at the moment, I didn't have time yet to port the most important piece ;)

**Comment by arichiardi** [comment]
Waiting for this on some feedback on #63 - the solution I have with dynamic vars is not really ideal imho

File: project.clj [hunk 1]
```diff
@@ -6,17 +6,31 @@
...
-                                  [fudje "0.9.7"]]
+                                  [fudje "0.9.7"]
+                                  [pjstadig/humane-test-output "0.9.0"]
+                                  ;; test-time dependencies
+                                  [org.clojure/core.async "0.4.490"]
+                                  [mount "0.1.15"]]
                    :plugins [[jonase/eastwood "0.3.5"]
-                             [lein-shell "0.5.0"]]
+                             [lein-shell "0.5.0"]
+                             [com.jakemccrary/lein-test-refresh "0.23.0"]]
```
Comment (alexander-yakushev) [inline project.clj]: Is this absolutely necessary for development?

File: project.clj [hunk 2]
```diff
@@ -6,17 +6,31 @@
   :profiles {:dev {:dependencies [[org.clojure/clojure "1.10.0"]
                                   [criterium "0.4.4"]
                                   [cloverage "1.0.13"]
-                                  [fudje "0.9.7"]]
+                                  [fudje "0.9.7"]
+                                  [pjstadig/humane-test-output "0.9.0"]
+                                  ;; test-time dependencies
+                                  [org.clojure/core.async "0.4.490"]
```
Comment (alexander-yakushev) [inline project.clj]: If I understand this correctly, these namespaces are needed only because the completion is tested against them. This is confusing. I would rewrite the tests so that they are tested against Clojure namespaces.

File: project.clj [hunk 3]
```diff
@@ -6,17 +6,31 @@
   :profiles {:dev {:dependencies [[org.clojure/clojure "1.10.0"]
                                   [criterium "0.4.4"]
                                   [cloverage "1.0.13"]
-                                  [fudje "0.9.7"]]
+                                  [fudje "0.9.7"]
+                                  [pjstadig/humane-test-output "0.9.0"]
```
Comment (alexander-yakushev) [inline project.clj]: This is orthogonal to this PR, please remove.

File: src/compliment/sources/cljs.cljc [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+       candidate)))
+
+(defn candidates
+  "Returns a sequence of candidate data for completions matching the given
+  prefix string and options in the ClojureScript compiler env.
+
+  - :cljs-env       - the ClojureScript compiler state;
+  - :context-ns     - (optional) namespace where completion is initiated as symbol;
+  - :extra-metadata - (todo) set of additional fields (:arglists, :doc) to add to the response maps."
+  [prefix {:keys [context-ns cljs-env] :as options-map}]   ;; WHAT TO DO HERE [^String prefix, ns context]
```
Comment (alexander-yakushev) [inline src/compliment/sources/cljs.cljc]: Regarding the dynamic variable, how ugly is ugly? I would expect it to be bound outside, and here you can rebind it to a lexical variable and then pass it around. Seems OKish to me, anything I'm missing?

File: src/compliment/utils.clj [hunk 5]
```diff
@@ -211,9 +211,31 @@
   []
   (let [classpath (classpath)]
     (cache-last-result ::project-resources classpath
-      (for [path classpath
-            ^String file (list-files path false)
-            :when (not (or (empty? file) (.endsWith file ".clj")
-                           (.endsWith file ".jar") (.endsWith file ".class")))]
-        (if (.startsWith file File/separator)
-          (.substring file 1) file)))))
+                       (for [path classpath
```
Comment (alexander-yakushev) [inline src/compliment/utils.clj]: Broken indentation.

File: project.clj (hunk 3 above)
Comment (arichiardi) [reply project.clj]: Ok will remove the dev tooling side of things at the end, it is just so useful at the moment 😁

File: project.clj (hunk 2 above)
Comment (arichiardi) [reply project.clj]: Some things were tested against core.async historically. Removing will take time so I will do at the end

File: src/compliment/sources/cljs.cljc (hunk 4 above)
Comment (arichiardi) [reply src/compliment/sources/cljs.cljc]: No nothing missing, maybe it is ugly to me only 😁

**Comment by arichiardi** [comment]
Ok pushed the version with the dynamic var.

I have also a question: given that Clojure and ClojureScript completions are mutually exclusive (for instance I get `or` in both `cljs.core` and `clojure.core`) I was wondering if there is a way to include only some completion sources and not others...

Or maybe cider should configure that from the outside

File: project.clj (hunk 3 above)
Comment (vemv) [reply project.clj]: Just as a handy tip, you can add these deps to your ~/.lein/profiles.clj profile so you can always have it in arbitrary projects

Comment (arichiardi) [reply project.clj]: Yeah no I know, I have just opened the conf from `cljs-tooling`. I will delete them, even though they improve the report and workflow in general. I generally dislike having globals in `profile.clj` and prefer to have self-contained projects...not a problem at all here - just explaining the rationale 😁

**Comment by arichiardi** [comment]
@alexander-yakushev during Clojure/North I got to a mostly usable version of the port.

I am missing still the `:doc` implementation but it is going to be very simple cause the compiler state basically contains it for most of the candidates.
I also probably don't strictly need it because Cider uses `orchard` for this.

In any case, maybe a second run of review would be good. Please remember that this is only 30% my code. Now I understand it completely, but the `cljs-tooling` code base was not really as optimized as this one.

**Comment by arichiardi** [comment]
Added the `doc` hook to the ClojureScript namespace today. The only TODO there is to return docstrings for macros. Small task.

A review now would be awesome.

**Comment by arichiardi** [comment]
pingy-ping :bell:

**Comment by alexander-yakushev** [comment]
I've skimmed through it. It looks pretty good!

> I have also a question: given that Clojure and ClojureScript completions are mutually exclusive (for instance I get or in both cljs.core and clojure.core) I was wondering if there is a way to include only some completion sources and not others...

I've needed this before as well but there is yet no way to do it. Is it necessary for this PR? If not, I might be able to think about it again later and maybe come up with something.

**Comment by alexander-yakushev** [comment]
How the tests should be run with these changes? When I plainly run `lein test` on your branch, it fails with `Caused by: java.io.FileNotFoundException: Could not locate cljs/env__init.class, cljs/env.clj or cljs/env.cljc on classpath.`

**Comment by arichiardi** [comment]
Yes you need to add the profile (in `cider` they use a Makefile for this).

Try:

```
 lein with-profile +1.10 test
```

**Comment by alexander-yakushev** [comment]
Sorry, missed that. Yes, that works.

**Comment by arichiardi** [comment]
I was thinking really that if the goal is to have the exact same api (I looove fuzzy completions) it might be maybe a better idea to include the correct broken-down Cljs implementations inside the current ones.

Meaning, a conditional on the presence of the dynamic var `cljs-env` could determine if ClojureScript completions are required.

Either that or to use the pieces of code from the Clojure completions that do fuzzy matching in the ClojureScript namespace. 

Does it make sense?

File: project.clj [hunk 6]
```diff
@@ -6,9 +6,11 @@
   :profiles {:dev {:dependencies [[org.clojure/clojure "1.10.0"]
                                   [criterium "0.4.4"]
                                   [cloverage "1.0.13"]
-                                  [fudje "0.9.7"]]
+                                  [fudje "0.9.7"]
+                                  [pjstadig/humane-test-output "0.9.0"]]
```
Comment (alexander-yakushev) [inline project.clj]: So, can you please remove those before we merge the PR?

File: project.clj [hunk 7]
```diff
@@ -17,6 +19,12 @@
                              "coverage" ["do" ["run" "-m" "compliment.t-coverage" "--coveralls"]
                                          ["shell" "curl" "-F"
                                           "json_file=@target/coverage/coveralls.json"
-                                          "https://coveralls.io/api/v1/jobs"]]}}
-             :1.8 {:dependencies [[org.clojure/clojure "1.8.0"]]}
-             :1.9 {:dependencies [[org.clojure/clojure "1.9.0"]]}})
+                                          "https://coveralls.io/api/v1/jobs"]]}
+                   :injections [(require 'pjstadig.humane-test-output)
+                                (pjstadig.humane-test-output/activate!)]}
```
Comment (alexander-yakushev) [inline project.clj]: Ditto.

File: src/compliment/utils.clj [hunk 8]
```diff
@@ -217,3 +217,39 @@
...
+
+(defn unquote-first
+  "Handles some weird double-quoting in the analyzer"
+  [form]
+  (if (= (first form) 'quote)
+    (first (rest form))
+    form))
+
+(defn normalize-arglists
+  [arglists]
```
Comment (alexander-yakushev) [inline src/compliment/utils.clj]: Could be useful to have a docstring for this function, otherwise it's hard to understand why exactly it is needed.

**Comment by arichiardi** [comment]
Comments addressed.

**Comment by alexander-yakushev** [comment]
Alright chums, let's do this!

**Comment by Olical** [comment]
😱🎉🎉🎉

I'll be trying this in Conjure ASAP!

**Comment by arichiardi** [comment]
Happy this is in now! But please let's keep in mind that I have a better (kinda breaking) improvement..

**Comment by arichiardi** [comment]
@alexander-yakushev it would be awesome to release a beta or snapshot version so that I can try it out in `cider-nrepl`

**Comment by alexander-yakushev** [comment]
Already did, `0.4.0-SNAPSHOT`.

**Comment by arichiardi** [comment]
Great thank you!
